src/library/scala/collection/immutable/Seq.scala

```diff
@@ -34,9 +34,54 @@ trait IndexedSeq[+A] extends Seq[A]
 
   final override def toIndexedSeq: IndexedSeq[A] = this
 
+  override def canEqual(that: Any): Boolean = that match {
+    case otherIndexedSeq: IndexedSeq[_] => length == otherIndexedSeq.length && super.canEqual(that)
+    case _ => super.canEqual (that)
+  }
+
+  protected def applyPreferredMaxLength: Int = IndexedSeqDefaults.defaultApplyPreferredMaxLength
+  override def equals(o: Any): Boolean = o match {
+    case that: IndexedSeq[_] =>
```

Comment:
As I noticed on the 2.12.x version of this PR, we still need to call `canEqual`.

Imagine we have a custom collection like `Range` that opts out of equality with other `Seq`-s in order to have a faster `hashCode` calculation.

```scala
final class MyRange(val start: Int, val end: Int) extends IndexedSeq[Int] {
  def apply(idx: Int) = start + idx
  def length = end - start
  override def canEqual(other: Any) = other.isInstanceOf[MyRange]
 override def equals(other: Any) = other match {
    case other: MyRange => other.start == start && other.end == end 
    case _ => false
  }
  override def hashCode = start + end
}
```

Reply:
We should add a unit test with this class to check equality is reflexive, ie, `assert(!MyRange(0, 1).equals(List(0)); assert(!List(0).equals(MyRange(0, 1)) `

Reply:
added unit test

Reply:
moved the inplementation to sameElements and added unit test